File: mediumSpringWebflux/src/main/java/com/example/mediumspringwebflux/global/config/SecurityConfig.java
package com.example.mediumspringwebflux.global.config;

import com.example.mediumspringwebflux.global.security.JwtWebFilter;
import com.example.mediumspringwebflux.global.security.Tokenizer;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.reactive.EnableWebFluxSecurity;
import org.springframework.security.config.web.server.SecurityWebFiltersOrder;
import org.springframework.security.config.web.server.ServerHttpSecurity;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.server.SecurityWebFilterChain;

@Configuration
@EnableWebFluxSecurity
public class SecurityConfig {

    private final Tokenizer tokenizer;

    public SecurityConfig(Tokenizer tokenizer) {
        this.tokenizer = tokenizer;
    }

    @Bean
    public PasswordEncoder passwordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Bean
    public SecurityWebFilterChain filterChain(ServerHttpSecurity http) {
        return http.authorizeExchange(exchanges -> exchanges
                .pathMatchers().permitAll()
                .anyExchange().authenticated())
                .csrf(ServerHttpSecurity.CsrfSpec::disable)
                .formLogin(ServerHttpSecurity.FormLoginSpec::disable)
                .addFilterBefore(new JwtWebFilter(tokenizer), SecurityWebFiltersOrder.HTTP_BASIC)
                .build();
    }

}

File: mediumSpringWebMVC/src/main/kotlin/com/example/mediumspringwebmvc/global/config/SecurityConfig.kt
package com.example.mediumspringwebmvc.global.config

import com.example.mediumspringwebmvc.global.security.JwtFilter
import com.example.mediumspringwebmvc.global.security.Tokenizer
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter

@Configuration
@EnableWebSecurity
class SecurityConfig(
    private val tokenizer: Tokenizer
) {

    @Bean
    fun filterChain(httpSecurity: HttpSecurity): SecurityFilterChain {
        httpSecurity.csrf { csrf -> csrf.disable() }
            .formLogin { formLogin -> formLogin.disable() }
            .headers { header -> header.frameOptions { frameOptions -> frameOptions.disable() } }
            .sessionManagement { session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .authorizeHttpRequests { auth -> auth
                .requestMatchers("/api/**").permitAll()
                .anyRequest().authenticated() }
            .addFilterBefore(JwtFilter(tokenizer), UsernamePasswordAuthenticationFilter::class.java)
        return httpSecurity.build()
    }


}